evap/results/views.py

```diff
@@ -1,59 +1,83 @@
+    if isinstance(courses, QuerySet):
+        courses = (courses
+            .annotate(num_participants=Count("participants", distinct=True), num_voters=Count("voters", distinct=True))
+            .select_related("type")
+            .prefetch_related(
+                "degrees",
+                "semester",
+                Prefetch("contributions", queryset=Contribution.objects.filter(responsible=True).select_related("contributor"), to_attr="responsible_contributions")
+            )
+        )
+        for course in courses:
+            course.responsible_contributors = [contribution.contributor for contribution in course.responsible_contributions]
+    for course in courses:
+        if not course.is_single_result:
+            course.distribution = calculate_average_distribution(course)
+            course.avg_grade = distribution_to_grade(course.distribution)
+        else:
+            course.single_result_rating_result = get_single_result_rating_result(course)
+    return courses
 
-    courses = semester.course_set.filter(state__in=visible_states).prefetch_related("degrees")
 
+@internal_required
+def index(request):
+    semesters = Semester.get_all_with_published_unarchived_results()
+    courses = Course.objects.filter(semester__in=semesters, state='published')
     courses = [course for course in courses if course.can_user_see_course(request.user)]
 
-    for course in courses:
-        course.distribution = calculate_average_distribution(course)
-        course.avg_grade = distribution_to_grade(course.distribution)
-
-    CourseTuple = namedtuple('CourseTuple', ('courses', 'single_results'))
-
-    courses_by_degree = OrderedDict()
-    for degree in Degree.objects.all():
-        courses_by_degree[degree] = CourseTuple([], [])
-    for course in courses:
-        if course.is_single_result:
-            for degree in course.degrees.all():
-                question_result = collect_results(course).questionnaire_results[0].question_results[0]
-                courses_by_degree[degree].single_results.append((course, question_result))
-        else:
-            for degree in course.degrees.all():
-                courses_by_degree[degree].courses.append(course)
+    if request.user.is_reviewer:
+        additional_courses = Course.objects.filter(semester__in=semesters, state__in=['in_evaluation', 'evaluated', 'reviewed'])
+        courses += get_courses_with_prefetched_data(additional_courses)
 
-    template_data = dict(semester=semester, courses_by_degree=courses_by_degree)
-    return render(request, "results_semester_detail.html", template_data)
+    course_pks = [course.pk for course in courses]
+    degrees = Degree.objects.filter(courses__pk__in=course_pks).distinct()
+    course_types = CourseType.objects.filter(courses__pk__in=course_pks).distinct()
+    template_data = dict(
+        courses=courses,
+        degrees=degrees,
+        course_types=sorted(course_types, key=lambda course_type: course_type.name),
```

Comment:
Maybe do the ordering in the database? (-> order_by on the QuerySet)

Reply:
we would need to check the language first and then decide to order on which column. here we're using the translated name already. while possible i think this doesn't really improve the current situation. what do you think?